kolibri/plugins/coach/assets/src/composables/useExerciseResources.js

```diff
@@ -0,0 +1,213 @@
+      ContentNodeResource.fetchDescendantsAssessments(results.map(topic => topic.id))
+        .then(({ data: topicsHasDescendantCounts }) => {
+          // Just get the IDs of the topics that have descendant exercises
+          const topicIdsWithDescendantCounts = topicsHasDescendantCounts
+            .filter(topic => topic.num_assessments > 0)
+            .map(topic => topic.id);
+
+          // Now, filter the topicResults to only include those topics
+          const topicsToAnnotate = topicResults.filter(topic =>
+            topicIdsWithDescendantCounts.includes(topic.id)
+          );
+
+          // Return those topics annotated to include the `num_assessments` property
+          return topicsToAnnotate.map(topic => {
+            topic.num_assessments = topicsHasDescendantCounts.find(
+              t => t.id === topic.id
+            ).num_assessments;
+            return topic;
+          });
+        })
+        .then(topicsWithCounts => {
+          const currentTopicContentList = topicsWithCounts.concat(exerciseResults);
+          if (moreParams) {
+            set(_contentList, [...get(_contentList), ...currentTopicContentList]);
+          } else {
+            set(_contentList, currentTopicContentList);
+          }
+        })
+        .then(() => set(_loading, false))
+        .catch(e => {
+          // TODO Work out best UX for this situation -- it may depend on if we're fetching more
+          // or the initial list of contents
+          console.error(e);
+          set(_loading, false);
+        });
+    });
+  }
+
+  /**
+   * @param {string} topicId The topicId for which we want to fetch the descendants
+   */
+  function setCurrentTopicId(topicId) {
+    if (!topicId) {
+      set(_currentTopicId, null);
+      set(_contentList, get(channels));
+    } else {
+      set(_currentTopicId, topicId);
+      fetchContentList();
+    }
+  }
+
+  /** @returns {Promise} */
+  function fetchChannelResources() {
+    return ChannelResource.fetchCollection({
+      params: { has_exercises: true, available: true },
+    }).then(response => {
+      console.log('setting channels', response);
+      set(
+        channels,
+        response.map(chnl => {
+          return {
+            ...chnl,
+            id: chnl.root,
+            title: chnl.name,
+            kind: ContentNodeKinds.CHANNEL,
+            is_leaf: false,
+          };
+        })
+      );
+      console.log('setted channels', get(channels));
+    });
+  }
+
+  /** @returns {Promise} */
+  function fetchBookmarks() {
+    return ContentNodeResource.fetchBookmarks({ params: { limit: 25, available: true } }).then(
+      data => {
+        bookmarks.value = data.results ? data.results : [];
+      }
+    );
+  }
+
+  function initializeExerciseResources() {
+    console.log('initializing resources');
+    set(_loading, true);
+    Promise.all([fetchChannelResources(), fetchBookmarks()])
+      .then(() => setCurrentTopicId(null))
```

Comment:
This line is currently causing a race condition when the page is first loaded - if this gets set after the topic id gets set from the route, then this will cause weird rendering.

With the computed prop for currentTopicId based on the route that I suggest above, this line can be removed, and will help prevent this problem.